service/collector/nat.go

```diff
@@ -0,0 +1,179 @@
+	"golang.org/x/sync/errgroup"
+
+	clientaws "github.com/giantswarm/aws-operator/client/aws"
+	"github.com/giantswarm/microerror"
+	"github.com/giantswarm/micrologger"
+)
+
+const (
+	labelVPC = "vpc"
+	labelAZ  = "availability_zone"
+)
+
+const (
+	subsystemNAT = "nat"
+)
+
+var (
+	natDesc *prometheus.Desc = prometheus.NewDesc(
+		prometheus.BuildFQName(namespace, subsystemNAT, "info"),
+		"NAT limits information.",
+		[]string{
+			labelAccountID,
+			labelVPC,
+			labelAZ,
+		},
+		nil,
+	)
+)
+
+type NATConfig struct {
+	Helper *helper
+	Logger micrologger.Logger
+
+	InstallationName string
+}
+
+type NAT struct {
+	helper *helper
+	logger micrologger.Logger
+
+	installationName string
+}
+
+func NewNAT(config NATConfig) (*NAT, error) {
+	if config.Helper == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Helper must not be empty", config)
+	}
+	if config.Logger == nil {
+		return nil, microerror.Maskf(invalidConfigError, "%T.Logger must not be empty", config)
+	}
+
+	if config.InstallationName == "" {
+		return nil, microerror.Maskf(invalidConfigError, "%T.InstallationName must not be empty", config)
+	}
+
+	v := &NAT{
+		helper: config.Helper,
+		logger: config.Logger,
+
+		installationName: config.InstallationName,
+	}
+
+	return v, nil
+}
+
+func (v *NAT) Collect(ch chan<- prometheus.Metric) error {
+	awsClientsList, err := v.helper.GetAWSClients()
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	var g errgroup.Group
+
+	for _, item := range awsClientsList {
+		awsClients := item
+
+		g.Go(func() error {
+			err := v.collectForAccount(ch, awsClients)
+			if err != nil {
+				return microerror.Mask(err)
+			}
+
+			return nil
+		})
+	}
+
+	err = g.Wait()
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	return nil
+}
+
+func (v *NAT) Describe(ch chan<- *prometheus.Desc) error {
+	ch <- natDesc
+	return nil
+}
+
+func (v *NAT) collectForAccount(ch chan<- prometheus.Metric, awsClients clientaws.Clients) error {
+	accountID, err := v.helper.AWSAccountID(awsClients)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	iv := &ec2.DescribeVpcsInput{
+		Filters: []*ec2.Filter{
+			{
+				Name: aws.String("tag:giantswarm.io/organization"),
+				Values: []*string{
+					aws.String("giantswarm"),
+				},
+			},
+		},
+	}
+	rv, err := awsClients.EC2.DescribeVpcs(iv)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	for _, vpc := range rv.Vpcs {
+		in := &ec2.DescribeNatGatewaysInput{
+			Filter: []*ec2.Filter{
+				{
+					Name: aws.String("vpc-id"),
+					Values: []*string{
+						vpc.VpcId,
+					},
+				},
+			},
+		}
+
+		var azs map[string]float64
+		azs = make(map[string]float64)
+		rn, err := awsClients.EC2.DescribeNatGateways(in)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+		for _, nat := range rn.NatGateways {
+			is := &ec2.DescribeSubnetsInput{
+				SubnetIds: []*string{
+					nat.SubnetId,
+				},
+			}
+			rs, err := awsClients.EC2.DescribeSubnets(is)
```

Comment:
There is a lot of complexity and loops in here and my fear is that we put too much pressure on the AWS API. I already saw the ELB collector cause rate limit issues so we should check if this here is the most efficient way to do it. Alternatively we go with this solution and improve on performance in another iteration.

Reply:
Yeah I stated [here](https://github.com/giantswarm/giantswarm/issues/8593#issuecomment-594047393) same concern. Any hint can we optimize this?

Reply:
Not really ATM.

Reply:
What about implement a simple cache. These values vary rarely, so instead of query API each 5 min, we can cache for half an hour. WDYT?

Reply:
Sounds good. Maybe [the caching implementation in `operatorkit`](https://github.com/giantswarm/operatorkit/blob/748c399a1bba09ef0227617d1e313c2723e16c5f/controller/string_cache.go) can help you find out how to do it.

Reply:
If I go now with this, meanwhile I implement in a following PR, it will impact all clusters? or only ones in latest GS version?

Reply:
For now it will not impact anything because your merge to master is not rolled out anywhere. In the coming days someone will maybe tag the master branch and create a new release with the code version.

Reply:
ok then let's go with this first implementation and I will provide the optimization tomorrow. I have a draft for it, but I would like to go one after other